File: cs4211-hw1/adapters/java/Arithmetic.java
import java.math.BigInteger;

/** GIVEN arbitrary-precision arithmetic operations for the two modes. */
final class Arithmetic {
    private Arithmetic() {}

    private static final BigInteger TWO_31 = BigInteger.ONE.shiftLeft(31);
    private static final BigInteger TWO_32 = BigInteger.ONE.shiftLeft(32);

    static BigInteger wrapResult(BigInteger value, Model.ArithmeticMode arithmetic) {
        if (arithmetic == Model.ArithmeticMode.INT32) {
            return value.add(TWO_31).mod(TWO_32).subtract(TWO_31);
        }
        return value;
    }

    static BigInteger divide(BigInteger left, BigInteger right,
                             Model.ArithmeticMode arithmetic) {
        if (right.signum() == 0) throw new Model.Stuck("division by zero");
        return wrapResult(left.divide(right), arithmetic);
    }

    static BigInteger apply(Model.ArithmeticOperator operator, BigInteger left,
                            BigInteger right, Model.ArithmeticMode arithmetic) {
        switch (operator) {
            case ADD: return wrapResult(left.add(right), arithmetic);
            case SUBTRACT: return wrapResult(left.subtract(right), arithmetic);
            case MULTIPLY: return wrapResult(left.multiply(right), arithmetic);
            case DIVIDE: return divide(left, right, arithmetic);
            default: throw new IllegalArgumentException("unknown operator " + operator);
        }
    }
}

File: cs4211-hw1/submission/arithmetic.py
"""GIVEN arithmetic helpers shared by the semantic rules."""

from model import ArithmeticMode, ArithmeticOperator, Stuck


def wrap_result(value: int, arithmetic: ArithmeticMode) -> int:
    """Wrap an operation result in int32 mode; numerals themselves are not wrapped."""

    if arithmetic is ArithmeticMode.INT32:
        return ((value + 2 ** 31) % 2 ** 32) - 2 ** 31
    return value


def divide(left: int, right: int, arithmetic: ArithmeticMode) -> int:
    """Divide with truncation toward zero."""

    if right == 0:
        raise Stuck("division by zero")
    quotient = abs(left) // abs(right)
    if (left >= 0) != (right >= 0):
        quotient = -quotient
    return wrap_result(quotient, arithmetic)


def apply_arithmetic_operator(
    operator: ArithmeticOperator,
    left: int,
    right: int,
    arithmetic: ArithmeticMode,
) -> int:
    """Apply one arithmetic operator using the homework's arithmetic mode."""

    if operator is ArithmeticOperator.ADD:
        return wrap_result(left + right, arithmetic)
    if operator is ArithmeticOperator.SUBTRACT:
        return wrap_result(left - right, arithmetic)
    if operator is ArithmeticOperator.MULTIPLY:
        return wrap_result(left * right, arithmetic)
    if operator is ArithmeticOperator.DIVIDE:
        return divide(left, right, arithmetic)
    raise ValueError("unknown arithmetic operator %r" % operator)
